span: add MultiFrontier data structure

andyyang890 opened the pull request:
Informs #148110

---

**span: make frontierHeap type-safe**

This patch makes `frontierHeap` type-safe by replacing the usage of the
go standard library heap package with the cockroach copy of the heap
package that uses generics.

Release note: None

---

**heaputil: add package with helper functions for working with heaps** 

This patch adds a new `heaputil` containing helper functions for
working with heaps. As of now, it only contains a single `Valid`
function which returns whether a heap is a valid heap based on
the min-heap invariant defined by `heap.Interface`.

The reasoning for creating a new package instead of reusing the
existing `heap` package is that that package is intended to be
a type-safe copy of the standard library package with the same name.

Release note: None

---

**span: add MultiFrontier data structure**

This patch adds a new `MultiFrontier` data structure that can be used
when we want a frontier that's partitioned into sub-frontiers under
the hood based on a custom partitioner function.

Release note: None

andyyang890 committed:
span: add MultiFrontier data structure

This patch adds a new `MultiFrontier` data structure that can be used
when we want a frontier that's partitioned into sub-frontiers under
the hood based on a custom partitioner function.

Release note: None

andyyang890 committed:
span: make frontierHeap type-safe

This patch makes `frontierHeap` type-safe by replacing the usage of the
go standard library heap package with the cockroach copy of the heap
package that uses generics.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/150855)
<!-- Reviewable:end -->

aerfrei commented on pkg/util/span/multi_frontier.go:
```diff
@@ -0,0 +1,328 @@
+// MultiFrontier is a Frontier that partitions its span space into
+// multiple sub-frontiers using a given PartitionerFunc.
+type MultiFrontier[T comparable] struct {
+	partitioner PartitionerFunc[T]
```
What type T/value for partitioner should I be imaging when I read this? Is the partition function going to take a span and match it to a table with T = tableId/descriptorId or something like that?

I think it would be helpful to elaborate in code comments a little about that

andyyang890 replied:
For the changefeed use case it'll be table IDs like you said. This PR is generic on purpose but the follow-up PR that integrates this into the changefeed code will definitely make that clear.

aerfrei replied:
I hear you. I think the reason I started to find this confusing was in the context of the test where we had the "testingTripartitePartitioner" T was also spans. I think that test could use a comment or two as well.

I think changing the comment for PartitionerFunc to something like:

```
// A PartitionerFunc is a function that assigns a span to the
// sub-frontier (indexed by T) it will be tracked within.
// e.g. For a per table frontier, this function would assign a
// span to the table it is located in. 
```

Choosing a consistent word from frontier, subfrontier, partition could help readability. I think the example in the comment would help to clarify even while the structure remains generic.

Relatedly, Is there another context we're planning on using the MultiFrontier?

andyyang890 commented on pkg/util/span/multi_frontier.go:
```diff
@@ -0,0 +1,336 @@
+	defer f.mu.Unlock()
+
+	for _, sp := range spans {
+		partition, err := f.partitioner(sp)
```
discussed this offline and we'll consider it in a later version but it's not necessarily that useful because spans could split/merge, i added a TODO comment

andyyang890 commented on pkg/util/span/multi_frontier.go:
```diff
@@ -0,0 +1,336 @@
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	frontier, ok := f.mu.frontiers.get(partition)
```
discussed offline and this is fine because the contract for `Forward` on `Frontier` states that any spans that don't overlap the tracked set will be ignored

andyyang890 commented on pkg/util/span/multi_frontier.go:
```diff
@@ -0,0 +1,336 @@
+		return errors.AssertionFailedf("frontier for partition %v already exists", partition)
+	}
+	elem := &multiFrontierHeapElem[T]{frontier: frontier, partition: partition}
+	heap.Push(h, elem)
```
discussed offline, it was previously handled by the `Push` method, but i've moved it out into the `add` function per https://github.com/cockroachdb/cockroach/pull/150855#discussion_r2246296376

andyyang890 commented on pkg/util/span/multi_frontier.go:
```diff
@@ -0,0 +1,336 @@
+	f.mu.frontiers.clear()
+}
+
+// Entries implements Frontier.
```
discussed this offline and we're going to leave it as-is for now because it matches the current behavior of the `concurrentFrontier` (see also https://github.com/cockroachdb/cockroach/issues/151217)

andyyang890 committed:
heaputil: add package with helper functions for working with heaps

This patch adds a new `heaputil` containing helper functions for
working with heaps. As of now, it only contains a single `Valid`
function which returns whether a heap is a valid heap based on
the min-heap invariant defined by `heap.Interface`.

The reasoning for creating a new package instead of reusing the
existing `heap` package is that that package is intended to be
a type-safe copy of the standard library package with the same name.

Release note: None

andyyang890 commented on pkg/util/span/multi_frontier_test.go:
```diff
@@ -0,0 +1,349 @@
+
+var _ span.PartitionerFunc[int] = testingTripartitePartitioner
+
+func multiFrontierStr[P cmp.Ordered](f *span.MultiFrontier[P]) string {
```
It also sorts the keys for consistent outputs in tests

andyyang890 commented on pkg/util/span/multi_frontier_test.go:
```diff
@@ -0,0 +1,349 @@
+	}
+	require.Equal(t,
+		`1: {[{a-b}@0.000000001,0]}, 2: {[{e-f}@0.000000004,0]}, 3: {[{i-j}@0.000000009,0]}`,
+		sortStr(f.String()))
```
Wanted to test the actual `String()` function instead of the test helper

andyyang890 commented on pkg/util/span/multi_frontier.go:
```diff
@@ -0,0 +1,396 @@
+// multiFrontierHeapP is a heap of frontiers that are sorted by their
+// frontier timestamps with convenience methods so that frontiers can
+// be accessed by their partition. It is not safe for concurrent use.
+type multiFrontierHeapP[P comparable] struct {
```
renamed to `partitionedMultiFrontierHeap`
